lucene/analysis/common/src/java/org/apache/lucene/analysis/hunspell/WordStorage.java

```diff
@@ -0,0 +1,338 @@
+      if (!entry.equals(currentEntry)) {
+        if (currentEntry != null) {
+          if (entry.compareTo(currentEntry) < 0) {
+            throw new IllegalArgumentException("out of order: " + entry + " < " + currentEntry);
+          }
+          int pos = flushGroup();
+
+          commonPrefixLength = GeneratingSuggester.commonPrefix(currentEntry, entry);
+          ByteArrayDataInput in = new ByteArrayDataInput(wordData);
+          in.setPosition(pos);
+          for (int i = currentEntry.length() - 1; i >= commonPrefixLength; i--) {
+            char c = (char) in.readVInt();
+            assert c == currentEntry.charAt(i);
+            pos -= in.readVInt();
+            in.setPosition(pos);
+          }
+          commonPrefixPos = pos;
+        }
+        currentEntry = entry;
+      }
+
+      group.add(flags);
+      if (hasCustomMorphData) {
+        morphDataIDs.add(morphDataID);
+      }
+    }
+
+    private int flushGroup() throws IOException {
+      IntsRefBuilder currentOrds = new IntsRefBuilder();
+
+      boolean hasNonHidden = false;
+      for (char[] flags : group) {
+        if (!hasHiddenFlag(flags)) {
+          hasNonHidden = true;
+          break;
+        }
+      }
+
+      for (int i = 0; i < group.size(); i++) {
+        char[] flags = group.get(i);
+        if (hasNonHidden && hasHiddenFlag(flags)) {
+          continue;
+        }
+
+        currentOrds.append(flagEnumerator.add(flags));
+        if (hasCustomMorphData) {
+          currentOrds.append(morphDataIDs.get(i));
+        }
+      }
+
+      int lastPos = commonPrefixPos;
+      for (int i = commonPrefixLength; i < currentEntry.length() - 1; i++) {
+        int pos = dataWriter.getPosition();
+        ensureArraySize(0, false);
+        dataWriter.writeVInt(currentEntry.charAt(i));
+        dataWriter.writeVInt(pos - lastPos);
+        lastPos = pos;
+      }
+
+      int pos = dataWriter.getPosition();
+      int hash = Math.abs(currentEntry.hashCode() % hashTable.length);
+      int collision = hashTable[hash];
+      hashTable[hash] = collision == 0 ? pos : -pos;
+
+      if (++chainLengths[hash] > 20) {
+        throw new RuntimeException(
+            "Too many collisions, please report this to dev@lucene.apache.org");
+      }
+
+      ensureArraySize(currentOrds.length(), collision != 0);
+      dataWriter.writeVInt(currentEntry.charAt(currentEntry.length() - 1));
+      dataWriter.writeVInt(pos - lastPos);
+      IntSequenceOutputs.getSingleton().write(currentOrds.get(), dataWriter);
+      if (collision != 0) {
+        dataWriter.writeByte(collision < 0 ? (byte) 1 : 0);
+        dataWriter.writeVInt(pos - Math.abs(collision));
+      }
+
+      group.clear();
+      morphDataIDs.clear();
+      return pos;
+    }
+
+    private void ensureArraySize(int valueLength, boolean hasCollision) {
+      int pos = dataWriter.getPosition();
+      int maxEntrySize = 8 + 4 * (valueLength + 1) + (hasCollision ? 5 : 0);
```

Comment:
This math looks maybe a bit fragile -- could we instead just insert an `ArrayUtil.grow` every time we try to append a byte to `wordData`?  I think we are not to concerned with CPU cost to build a new dictionary right?  We build once and then `.get` zillions of times?

Hmm I thought Lucene had something like `ByteArrayDataOutput` that already did this under-the-hood, but I looked and I'm not sure it does!  We should add that!  I see `ByteArrayDataOutput.reset(...)` so caller could do that on the outside.

Reply:
I don't like this math either, but we don't write bytes, we write vInts, and they're trickier. And calling `grow` before each of them would make the code messier.

I wish `ByteArrayDataOutput` would be smarter indeed. I'm not sure though that all clients would be ready that the underlying `byte[]` can change, but I probably can override `writeByte` here locally. WDYT?

Reply:
Hi Peter. :) Sorry for absence, holidays, catching-up with work.

For a growable buffer you can use OutputStreamDataOutput with a ByteArrayOutputStream or (probably better?) ByteBuffersDataOutput and then call toArrayCopy which returns a contiguous array from internal chunked slices.

Reply:
Hi! Thanks, that's what I was looking for. Am I right that ByteBuffersDataOutput is better because it doesn't have to copy everything when growing?

Reply:
Precisely. It'll still collect into a contiguous array at the end (after which you can null the original reference and it can be gc'ed). I don't think it's a problem. 

You could also use the output of ByteBuffersDataOutput directly as a (seekable) datainput... but I don't know if this makes sense if you're trying to squeeze the last bits of performance.

Reply:
(but if you'd like to try then .toDataInput() return a seekable data input over the original slices of data).

Reply:
But it's not that easy: we also need access to the previously written bytes. AFAIU neither of your suggestions provide easy access to the underlying bytes?

Reply:
But toDataInput doesn't survive new output being written, right? Then I'd have to call it for roughly each entry, that's a lot of work :(

Reply:
Hmmm... I didn't realize you need access to those bytes while you're still writing them. In that case yes - it'll be a problem. You could use toDataInput - this would provide you with a view over previously written bytes. I don't know if it's worth the complication though. If it's so specific here, just oversize the local byte[] as it is.

Reply:
For the record - OutputStreamDataOutput with a ByteArrayOutputStream would work if you subclass ByteArrayOutputStream and expose the underlying buf... But again - it's no different than resizing manually.

Reply:
I am surprised we do not already have a simple and efficient class to write to chunks of growing `byte[]` and also provide an efficient view (`DataInput`) on those bytes as they are being written too ... hrmph.

Reply:
I think there used to be one but I removed it because it seemed like a duplicate of the possibilities I listed... it was growable... something?